internal/application/application.go

```diff
@@ -247,19 +248,92 @@ func List(ctx context.Context, cluster *kubernetes.Cluster, namespace string) (m
 		}
 	}
 
-	// Get references for all apps, deployed or not
+	// Fast batch queries to load all relevant resources in as few kube calls as possible.
 
-	appRefs, err := ListAppRefs(ctx, cluster, namespace)
+	// I. Get the application resources for all apps, deployed or not
+
+	client, err := cluster.ClientApp()
+	if err != nil {
+		return nil, err
+	}
+	appCRList, err := client.Namespace(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// II. Load the auxiliary application data found in adjacent kube Secret resources
+	//     (environment, scaling, bound configs).
+
+	secrets, err := cluster.Kubectl.CoreV1().Secrets(namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/managed-by=epinio",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// III. The pods for the deployed apps.
+
+	pods, err := ListApplicationPods(ctx, cluster, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// IV. Actual application routes from the ingresses
+
+	routes, err := ListActualApplicationRoutes(ctx, cluster, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// V. Pod metrics and replica information
+
+	metrics, err := GetPodMetrics(ctx, cluster, namespace)
 	if err != nil {
-		return models.AppList{}, err
+		return nil, err
 	}
 
-	// Convert references to full application structures
+	// Fuse the loaded resources into full application structures.
+	//
+	// Note: The returned secrets are a mix of scaling instructions, bound configurations, and
+	// environment assignments. Split them into separate maps as per their area (*). Key the
+	// maps by namespace and name of their controlling application for quick access in the
+	// aggregation step.
+	//
+	// (*) Label "epinio.io/area": "environment"|"scaling"|"configuration"
+
+	scales := make(map[string]v1.Secret)
+	bounds := make(map[string]v1.Secret)
+	environs := make(map[string]v1.Secret)
+
+	for _, s := range secrets.Items {
+		area, found := s.Labels["epinio.io/area"]
+		if !found {
+			continue
+		}
+		app, found := s.Labels["app.kubernetes.io/name"]
+		if !found {
+			continue
+		}
+
+		key := ConfigurationKey(app, s.ObjectMeta.Namespace)
+
+		switch area {
+		case "scaling":
+			scales[key] = s
+		case "configuration":
+			bounds[key] = s
+		case "environment":
+			environs[key] = s
+		default:
+			// ignore secret
+		}
+	}
```

Comment:
I would move this into a separate func, using a dedicated struct instead of multiple maps (`buildAppSecretsMap` or something like that):

```go
type appSecrets struct {
	scaling       *v1.Secret // pointer to check if nil
	configuration *v1.Secret
	environment   *v1.Secret
}

func buildAppSecretsMap(appName, namespace string, secrets secrets.Items) map[string]appSecrets {
	// do stuff
}
```

so we can have a `map[string]appSecrets` to match all the "settings" for an app.

Not sure if we could add also the other info there (pods, routes, metrics).

Reply:
That should be do-able. Even for the pods, etc. too.

Reply:
Well, the metrics are keyed by pod name, not by app+namespace as the rest. Thus metrics stay as a separate map.

Done with commit ce4db9642c7cf87e4ea7a1620758276a9e34557f